nameserver/cache.go

```diff
@@ -0,0 +1,318 @@
+package nameserver
+
+import (
+	"errors"
+	"github.com/miekg/dns"
+	"math"
+	"sort"
+	"sync"
+	"time"
+	//"fmt"
+	"math/rand"
+)
+
+var (
+	errInvalidCapacity = errors.New("Invalid cache capacity")
+	errCouldNotResolve = errors.New("Could not resolve")
+	errTimeout         = errors.New("Timeout while waiting for resolution")
+)
+
+const (
+	defPendingTimeout int = 5 // timeout for a resolution
+)
+
+type entryStatus uint8
+
+const (
+	stPending  entryStatus = iota // someone is waiting for the resolution
+	stResolved entryStatus = iota // resolved
+)
+
+const (
+	CacheLocalReply uint8 = 1 << iota // the reply was obtained from a local resolution
+)
+
+// shuffleAnswers reorders answers for very basic load balancing
+func shuffleAnswers(answers []dns.RR) []dns.RR {
+	rand.Seed(time.Now().UTC().UnixNano())
+
+	n := len(answers)
+	for i := 0; i < n; i++ {
+		r := i + rand.Intn(n-i)
+		answers[r], answers[i] = answers[i], answers[r]
+	}
+
+	return answers
+}
+
+// a cache entry
+type cacheEntry struct {
+	Status entryStatus // status of the entry
+	Flags  uint8       // some extra flags
+
+	question dns.Question
+	protocol dnsProtocol
+	reply    dns.Msg
+
+	validUntil time.Time // obtained from the reply and stored here for convenience/speed
+	putTime    time.Time
+
+	waitChan chan struct{}
+}
+
+func newCacheEntry(question *dns.Question, reply *dns.Msg, status entryStatus, now time.Time) *cacheEntry {
+	e := &cacheEntry{
+		Status:   status,
+		Flags:    0,
+		question: *question,
+	}
+
+	if e.Status == stPending {
+		e.validUntil = now.Add(time.Duration(defPendingTimeout) * time.Second)
+		e.waitChan = make(chan struct{})
+	} else {
+		e.setReply(reply, 0, now)
+	}
+
+	return e
+}
+
+// Get a copy of the reply stored in the entry, but with some values adjusted like the TTL
+func (e *cacheEntry) getReply(request *dns.Msg, now time.Time) (*dns.Msg, error) {
+	if e.Status != stResolved {
+		return nil, nil
+	}
+
+	// if the reply has expired or is invalid, force the caller to start a new resolution
+	if e.hasExpired(now) {
+		return nil, nil
+	}
+
+	// create a copy of the reply, with values for this particular query
+	reply := e.reply
+	reply.SetReply(request)
+	reply.Rcode = e.reply.Rcode
+	reply.Authoritative = (e.Flags&CacheLocalReply != 0) // we are only authoritative for local questions
+
+	// adjust the TTLs
+	passedSecs := uint32(now.Sub(e.putTime).Seconds())
```

Comment:
I have taken the conservative path: an entry is valid until all the RR in the response are valid. I was specially concerned about _additional_ records in the response that could increase the `validUntil` if it was calculated that way.

For example, if asking about `some.domain.com NS` you could get the list of nameservers (`some.domain.com. 21223 IN NS ns1.some.domain.com.`) and, additionally, the list of IPs for those nameservers (`ns1.some.domain.com. 559 IN A 144.168.1.100.`), both with different TTLs, but both the nameservers and the names would have to be up-to-date in order to be useful for the client. Otherwise, the client would bother us again with another query, it would take more space from the cache...